cylc/uiserver/schema.py

```diff
@@ -396,78 +415,299 @@ def run_task_query(conn, workflow):
+        'waiting'
+
+        >>> _status_to_state(_state_to_status(0, 1, True))
+        (0, 1, True)
+
+    """
+    if run_status is not None:
+        if run_status == 0:
+            status = TASK_STATUS_SUCCEEDED
+        else:
+            status = TASK_STATUS_FAILED
+    elif time_run is not None:
+        status = TASK_STATUS_RUNNING
+    elif submit_status is not None:
+        if submit_status == 0:
+            status = TASK_STATUS_SUBMITTED
+        else:
+            status = TASK_STATUS_SUBMIT_FAILED
+    else:
+        status = TASK_STATUS_WAITING
 
-    # TODO: support all arguments including states
-    # https://github.com/cylc/cylc-uiserver/issues/440
+    return status
+
+
+def run_jobs_query(
+    conn,
+    workflow,
+    ids=None,
+    exids=None,
+    states=None,
+    exstates=None,
+    tasks=None,
+):
+    """Query jobs from the database.
+
+    Args:
+        conn: Database connection.
+        workflow: Workflow ID.
+        kwargs: GraphQL sort/filter args as per cylc-flow interfaces.
+
+    """
+    # TODO: support all arguments:
+    # * [x] ids
+    # * [ ] sort
+    # * [x] exids
+    # * [x] states
+    # * [x] exstates
+    # See https://github.com/cylc/cylc-uiserver/issues/440
     jobs = []
+    where_stmts = []
+    where_args = []
+
+    # filter by cycle/task/job ID
+    if ids:
+        items = []
+        for id_ in ids:
+            item = []
+            for token, column in (
+                ('cycle', 'cycle'),
+                ('task', 'name'),
+                ('job', 'submit_num'),
+            ):
+                value = id_[token]
+                if value:
+                    if token == 'job':
+                        value = int(value)
+                    item.append(rf'{column} GLOB ?')
+                    where_args.append(value)
+            items.append(r'(' + ' AND '.join(item) + r')')
+
+        if items:
+            where_stmts.append(
+                r'(' + ' OR '.join(items) + ')'
+            )
+
+    # filter out cycle/task/job IDs
+    if exids:
+        for id_ in exids:
+            items = []
+            for token, column in (
+                ('cycle', 'cycle'),
+                ('task', 'name'),
+                ('job', 'submit_num'),
+            ):
+                value = id_[token]
+                if value:
+                    if token == 'job':
+                        value = int(value)
+                    items.append(rf'{column} GLOB ?')
+                    where_args.append(value)
+            if items:
+                where_stmts.append(r'NOT (' + ' AND '.join(items) + r')')
+
+    # filter by job state
+    if states:
+        items = []
+        for status in states:
+            submit_status, run_status, time_run = _status_to_state(status)
+            if submit_status is None:
+                # hasn't yet submitted (i.e. there is no job)
+                continue
+            item = [r'IFNULL(submit_status,999) = ?']
```

Comment:
FYI: SQL skips comparisons for Null values, so we need to use `IFNULL` to force the comparison to be respected.